### DISCUSSION TITLE
About Middleware

### DISCUSSION THREAD

**Discussion by josejachuf** [salvo-rs/salvo#155] [General]
Hi
With the Middleware can something similar to this [1] be achieved?
It can be manipulated before and after the handler
Jose

[1] https://emmett.sh/docs/2.4.x/pipeline

**Comment by chrislearn** [comment]
When router matched, handlers will collected in a Vec.
Salvo will call the first Handler's handle function. This handler will call next handle...and so on.

```rust
#[async_trait]
impl Handler for Middleware {
    async fn handle(&self, req: &mut Request, depot: &mut Depot, res: &mut Response, ctrl: &mut FlowCtrl) {
        ctrl.call_next(req, depot, res).await;
    }
}
```
```ctrl.call_next(req, depot, res)``` will call next handler.
If you want before handler,  You can:
```rust
 async fn handle(&self, req: &mut Request, depot: &mut Depot, res: &mut Response, ctrl: &mut FlowCtrl) {
      // Write your logic code here
      ctrl.call_next(req, depot, res).await;
  }
```
https://github.com/salvo-rs/salvo/blob/main/extra/src/size_limiter.rs

Or after handler:
```rust
 async fn handle(&self, req: &mut Request, depot: &mut Depot, res: &mut Response, ctrl: &mut FlowCtrl) {
      ctrl.call_next(req, depot, res).await;
      // Write your logic code here
  }
```
https://github.com/salvo-rs/salvo/blob/main/extra/src/compression.rs

**Reply by josejachuf** [reply]
Thanks. I will see it
